### PR TITLE
An update to #15582: clarify remaining obscure points regarding decompose_{lambda,prod}{_n,}{_assum,_decls} functions

### PR DESCRIPTION
**Pull request by herbelin**
Issue #15562 and its implementation in #15582 left a couple of inconsistencies in decompose_{lambda,prod}{_n,}{_assum,_decls} functions (see the August 2023 addendum to #15562).

After #15582, the following inconsistencies remain:
- Term.decompose_lambda_n_assum (called Term.decompose_lam_n_assum before 15582) started to fail on let-ins, but not EConstr.decompose_lambda_n_assum (called EConstr.decompose_lam_n_assum before)
- Reduction.hnf_decompose_lambda_n_decls (called Reduction.dest_lam_n_assum before) was preserving (and not counting) let-ins but Reductionops.hnf_decompose_prod_n_assum (called Reductionops.splay_prod_n before) was not
- Reductionops.hnf_decompose_lambda_n_assum was not preserving let-ins
- some of the hnf_decompose_{lambda,prod}* functions had a heuristic to deal with let-ins blocking redexes but not all
- some related functions were omitted from the cleanup such as Reductionops.dest_prod_assum whis was essentially the same as Reductionops.hnf_decompose_prod_decls

Also, issue #15562 questioned the arbitrariness of what to do with trailing let-ins when decomposing a block of binders by counting only the Lambda. Besides the non-uniform changes of semantics done in #15582, this also overlooked that the behavior consisting in not counting let-ins but preserving them is exactly what is needed to extract the initial context of a `Fix` knowing the index of the decreasing argument. So, the preserving but not counting LetIn semantics seems better.

The PR addresses all the above discrepancies:
- the \_assum suffix to decompose_{lambda,prod} now always preserves let-ins but not counting them, and the hnf_ variants always use a heuristics for let-ins blocking redexes
- the _decls suffix always preserve and count let-ins

In particular:
- it deprecates `Reductionops.hnf_decompose_prod_n_decls` and `Reductionops.hnf_decompose_lambda_n_assum` (which were not preserving let-ins)
- In `reductionops.ml`, it adds instead `hnf_decompose_{lambda,prod}_n` (that does not preserve let-ins) and `hnf_decompose_prod_n_assum`, `New.hnf_decompose_prod_n_decls`, `New.hnf_decompose_lambda_n_assum` (that do preserve let-ins)
- it aligns `Term.decompose_lambda_n_assum` on `EConstr.decompose_lambda_n_assum` (preserving but not counting let-ins)
- it renames `Reduction.hnf_decompose_lambda_n_decls` to `Reduction.hnf_decompose_lambda_n_assum` to reflect its actual semantics of preserving but not counting let-ins

The PR additionally adds some documentation of these functions.

### CODE REVIEW

**Comment by ppedrot** [comment]
> Term.decompose_lam_n_assum is apparently not used in the Coq archive!

This, coupled with the fact that there is some wiggle room in the implementation when compared to the spec that this PR is fixing, would rather suggest that we should not expose this function...

**Comment by herbelin** [comment]
Actually, I realized the change of definition precisely when looking for "the" function that extracts the context of the body of a fix exactly up to the decreasing argument (I use such function in an update of paramcoq that iota-expands `fix`s - Cyril's trick - to avoid having to prove manually that `fix` are observationally equal to their expansion).

**Comment by herbelin** [comment]
@ppedrot: I fixed other discrepancies regarding the _assum and _decls functions.

Regarding `Term.decompose_lambda_n_assum`, #15453 reminded me that the right function to use in fixpoints is actually what the PR names `Reduction.hnf_decompose_lambda_n_assum` (doing head-normalization). So, indeed, `Term.decompose_lambda_n_assum` could be removed.

**Comment by ppedrot** [comment]
By the way, I realized after #15582 that it probably picked a confusing scheme for `hnf_`. Indeed, `hnf` is also the name of the user-facing reduction tactic which is really not at all doing what OCaml hnf functions are doing, since it's using a variant of `simpl`. As a matter of fact, we also have the `hnf_constr` OCaml function which behaves like the user `hnf` function, and that one settled the name scheme. This keeps confusing me again and again.

Consequently I think that instead of `hnf_`, the newly renamed functions should follow the kernel convention for reduction and use `whd_` instead...

**Comment by herbelin** [comment]
Yes, `whd_` looks better to me too. Do we dare another renaming?

File: kernel/reduction.mli [hunk 1]
```diff
@@ -66,4 +71,6 @@ val dest_prod_assum : env -> types -> Constr.rel_context * types
 val dest_lam        : env -> constr -> Constr.rel_context * constr
 [@@ocaml.deprecated "Use [hnf_decompose_lambda] instead."]
 val dest_lam_assum  : env -> constr -> Constr.rel_context * constr
-[@@ocaml.deprecated "Use [hnf_decompose_lambda_decls] instead."]
+[@@ocaml.deprecated "Use [hnf_decompose_lambda_assum] instead."]
+val hnf_decompose_lambda_n_decls : env -> int -> constr -> Constr.rel_context * constr
+[@@ocaml.deprecated "Use [hnf_decompose_lambda_n_assum] instead."]
```
[the remark this answers is not in the document]
Comment (herbelin) [reply kernel/reduction.mli:76]: I'm not so sure. For instance, when I worked on paramcoq, the deprecation notices were useful to know in one second how to adapt to the API. Also, from time to time, I seize the opportunity of running the CI to take the warnings of plugins into account.

It is not really a problem either if the deprecation notice stays for a few years.

What I find is a problem though is the proliferation of warnings: when there are too many of them (the noise you are talking about probably), we don't know with which to start.

[the remark this answers is not in the document]
Comment (JasonGross) [reply kernel/reduction.mli]: FWIW I agree with @herbelin here.  The biggest benefit of deprecation warnings is not compatibility but the usefulness of having at least one version of Coq that will tell you immediately how to fix the issue.

**Comment by ppedrot** [comment]
@coqbot run full ci

**Comment by ppedrot** [comment]
@herbelin can you rebase to get a clean CI?

**Comment by herbelin** [comment]
@coqbot run full ci

**Comment by herbelin** [comment]
CI still not clean:
- vscoq missing the merge of coq-community/vscoq#654
- 2 timeouts

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@coqbot merge now

**Pull request merged**